pkg/ring/kv/multi.go

```diff
@@ -0,0 +1,349 @@
+	f.Float64Var(&cfg.MirrorRateLimit, prefix+"multi.mirror-rate-limit", 1, "Rate limit used for mirroring. 0 disables rate limit.")
+	f.IntVar(&cfg.MirrorRateBurst, prefix+"multi.mirror-burst-size", 1, "Burst size used by rate limiter. Values less than 1 are treated as 1.")
+}
+
+// MultiRuntimeConfig has values that can change in runtime (via overrides)
+type MultiRuntimeConfig struct {
+	// Primary store used by MultiClient. Can be updated in runtime to switch to a different store (eg. consul -> etcd,
+	// or to gossip). Doing this allows nice migration between stores. Empty values are ignored.
+	PrimaryStore string `yaml:"primary"`
+
+	// Mirroring enabled or not. Nil = no change.
+	Mirroring *bool `yaml:"mirroring"`
+}
+
+type kvclient struct {
+	client Client
+	name   string
+}
+
+type clientInProgress struct {
+	client int
+	cancel context.CancelFunc
+}
+
+// MultiClient implements kv.Client by forwarding all API calls to primary client.
+// At the same time, MultiClient watches for changes in values in the primary store,
+// and forwards them to remaining clients.
+type MultiClient struct {
+	// Available KV clients
+	clients []kvclient
+
+	rateLimiter      *rate.Limiter
+	mirroringEnabled *atomic.Bool
+
+	// Primary client used for interaction. Values from this KV are copied to all the others.
+	primaryID *atomic.Int32
+
+	cancel context.CancelFunc
+
+	inProgressMu sync.Mutex
+	// Cancel functions for ongoing operations. key is a value from inProgressCnt.
+	// What we really need is a []context.CancelFunc, but functions cannot be compared against each other using ==,
+	// so we use this map instead.
+	inProgress    map[int]clientInProgress
+	inProgressCnt int
+}
+
+// NewMultiClient creates new MultiClient with given KV Clients.
+// First client in the slice is the primary client.
+// Channel is used to get notifications about what store to use as primary.
+func NewMultiClient(cfg MultiConfig, clients []kvclient) *MultiClient {
+	c := &MultiClient{
+		clients:    clients,
+		primaryID:  atomic.NewInt32(0),
+		inProgress: map[int]clientInProgress{},
+
+		rateLimiter:      createRateLimiter(cfg.MirrorRateLimit, cfg.MirrorRateBurst),
+		mirroringEnabled: atomic.NewBool(cfg.MirrorEnabled),
+	}
+
+	ctx, cancelFn := context.WithCancel(context.Background())
+	c.cancel = cancelFn
+
+	// Start mirroring.
+	go c.mirrorChanges(ctx, cfg.MirrorPrefix)
+
+	if cfg.ConfigProvider != nil {
+		go c.watchConfigChannel(ctx, cfg.ConfigProvider())
+	}
+
+	return c
+}
+
+func (m *MultiClient) watchConfigChannel(ctx context.Context, configChannel <-chan MultiRuntimeConfig) {
+	for {
+		select {
+		case cfg, ok := <-configChannel:
+			if !ok {
+				return
+			}
+
+			if cfg.Mirroring != nil {
+				enable := *cfg.Mirroring
+				old := m.mirroringEnabled.Swap(enable)
+				if old != enable {
+					level.Info(util.Logger).Log("msg", "toggled mirroring", "enabled", enable)
+				}
+			}
+
+			if cfg.PrimaryStore != "" {
```

Comment:
This should never happen. Maybe log an error?

Reply:
It will happen if primary store is not defined in overrides.yaml. This entire method is dealing with values loaded from overrides.